compiler/src/iree/compiler/Codegen/Common/EncodingUtils.cpp

```diff
@@ -5,46 +5,125 @@
+      tensorType.getEncoding().dyn_cast_or_null<EncodingAttr>();
+  if (!encoding || tensorType.getRank() <= 1) {
+    return tensorType;
+  }
+  if (!isNarrowNResult(encoding)) {
+    return tensorType;
+  }
+  if (tensorType.getRank() <= 1) {
+    return tensorType;
+  }
+  auto newRole = encoding.getRole().getValue();
+  TypeAttr originalTypeAttr = encoding.getOriginalType();
+  RankedTensorType originalType;
+  if (originalTypeAttr) {
+    originalType = originalTypeAttr.getValue().dyn_cast<RankedTensorType>();
+  }
+  if (!originalType) {
+    originalType = tensorType;
+  }
+  SmallVector<int64_t> newOriginalShape{originalType.getShape()};
+  auto userIndexingMaps = encoding.getUserIndexingMaps();
+  SmallVector<AffineMap> maps;
+  for (auto a : userIndexingMaps) {
+    maps.push_back(cast<AffineMapAttr>(a).getAffineMap());
+  }
+  auto cDims = linalg::inferContractionDims(maps);
+  SmallVector<int64_t> newShape{tensorType.getShape()};
+  SmallVector<int64_t> permIndices(maps[0].getNumDims());
+  for (int i = 0; i < permIndices.size(); ++i) {
+    permIndices[i] = i;
+  }
+  // Matrix case: there are both M and N dimensions. Transposing means swapping
+  // them.
+  if (cDims->m.size() == 1 && cDims->n.size() == 1) {
+    int m = cDims->m[0];
+    int n = cDims->n[0];
+    std::swap(newShape[m], newShape[n]);
+    std::swap(newOriginalShape[m], newOriginalShape[n]);
+    std::swap(permIndices[m], permIndices[n]);
+  }
+  // Vector case: there is no N dimension to swap the M dimension with. We
+  // swap the maps themselves.
+  if (cDims->n.empty()) {
+    std::swap(maps[0], maps[1]);
+  }
+  AffineMap permutation =
+      AffineMap::getPermutationMap(permIndices, tensorType.getContext());
+  for (auto &map : maps) {
+    map = map.compose(permutation);
+  }
+  SmallVector<Attribute> newMaps;
+  for (auto map : maps) {
+    newMaps.push_back(AffineMapAttr::get(map));
+  }
+  ArrayAttr newIndexingMaps = ArrayAttr::get(tensorType.getContext(), newMaps);
+  auto newEncoding = EncodingAttr::get(
+      encoding.getContext(),
+      IREE::LinalgExt::EncodingRoleAttr::get(encoding.getContext(), newRole),
+      encoding.getElementTypes(),
+      TypeAttr::get(
+          RankedTensorType::get(newOriginalShape, tensorType.getElementType())),
+      encoding.getMatmulNarrow_N(), encoding.getMatmulNarrow_M(),
+      newIndexingMaps);
+  auto r =
+      RankedTensorType::get(newShape, tensorType.getElementType(), newEncoding);
+  return r;
+}
+
 /// For a given tensor type with an encoding, return the materialized
 /// type to use for it. If no encoding is set, then return the tensor type
 /// itself.
 static RankedTensorType
 getMaterializedType(RankedTensorType tensorType,
-                    MaterializeEncodingFn materializeEncodingFn) {
+                    MaterializeEncodingFn materializeEncodingFn,
+                    bool transposeNarrowN) {
+  if (transposeNarrowN) {
+    tensorType = transposeIfNarrowNResult(tensorType);
+  }
```

Comment:
Do we need the check? I think `transposeIfNarrowNResult` already does it for us. In this context, we don't need to pass `transposeNarrowN` to the function either.

Reply:
Moved to the caller, so we can drop the `transposeNarrowN` parameter. But the check itself is not redundant. `transposeIfNarrowNResult` can't perform this check, it doesn't even get `transposeNarrowN` passed to it.